plugins/inputs/powerdns_recursor/protocol_commons.go

```diff
@@ -0,0 +1,96 @@
+package powerdns_recursor
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"unsafe"
+)
+
+func parseResponse(metrics string) map[string]interface{} {
+	values := make(map[string]interface{})
+
+	s := strings.Split(metrics, "\n")
+
+	for _, metric := range s[:len(s)-1] {
+		m := strings.Split(metric, "\t")
+		if len(m) < 2 {
+			continue
+		}
+
+		i, err := strconv.ParseInt(m[1], 10, 64)
+		if err != nil {
+			continue
+		}
+
+		values[m[0]] = i
+	}
+
+	return values
+}
+
+// This below is generally unsafe but necessary in this case
+// since the powerdns protocol encoding is host dependent.
+// The C implementation uses size_t as the size type for the
+// command length. The size and endianness of size_t change
+// depending on the platform the program is being run on.
+// At the time of writing, the Go type `uint` has the same
+// behavior, where its size and endianness are platform
+// dependent. Using the unsafe method below, and the known
+// integer size, we can "recreate" the corresponding C
+// behavior in an effort to maintain compatibility. Of course
+// in cases where one program is compiled for i386 and the
+// other for amd64 (and similar), this method will fail.
+
+const uintSizeInBytes = strconv.IntSize / 8
+
+func getEndianness() binary.ByteOrder {
+	buf := make([]byte, 2)
+	*(*uint16)(unsafe.Pointer(&buf[0])) = uint16(0x0001)
+
+	if buf[0] == 1 {
+		return binary.LittleEndian
+	}
+
+	return binary.BigEndian
+}
```

Comment:
Alternatively, we can make this more safe by defining the `getEndianess` function depending on the host. See the [binary parser's host_endianess_*.go files](https://github.com/influxdata/telegraf/tree/master/plugins/parsers/binary). What do you think? Maybe we should generalize this to a `internal`'s function.

Reply:
I did consider having target-specific implementations for this, but I ended up choosing to do it like this in order to avoid bugs from new target architectures being introduced. The example you sent for instance is already broken, since it only returns big endian for `arm64be`, while go also has support for other big-endian architectures (like `ppc` and `mips`).

Reply:
Right, but this can (and should) be fixed... @powersj what do you think?

Reply:
I would like for us to avoid using unsafe and instead by very explicit about what build architectures use which. When something is missing it is straightforward to add and clear what would use what

Looking through [the valid GOARCH values](https://gist.github.com/lizkes/975ab2d1b5f9d5fdee5d3fa665bcfde6#a-list-of-valid-goarch-values) of the arches we actually build binaries for, does this list capture the current big-endian architectures?

* mips
* mips64
* ppc64
* s390x

Reply:
Actual source of truth for supported architectures should be [here](https://github.com/golang/go/blob/master/src/go/build/syslist.go). I believe from that list the following are big endian:
- armbe
- arm64be
- mips
- mips64
- mips64p32
- ppc
- ppc64
- s390
- s390x
- sparc
- sparc64

Reply:
Ah I didn't even realize I linked to the  repo! thanks 

your list will work

Reply:
hey - would like to see this land in v1.25 on Monday, would you be willing to push an update before then?

Reply:
Oh, I was under the impression you'd be taking over from here to introduce some internal way to get host endianness so it can be used in the binary parser as well. I'm happy to either adjust this plugin specifically or attempt to expose it internally, but I'm not sure if I'll find time before Monday.

Reply:
@OrfeasZ and @powersj I quickly put together a PR #12360. Happy to get your feedback!

Reply:
@OrfeasZ the PR is merged, can you please update your PR to use the newly added declaration!?

Reply:
Done!

Reply:
@srebhan I also just double checked that compilation fails for unspecified architectures, and everything seems to work as expected. Should expect to see variable `undefined` errors when that happens.